tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,457 @@
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
+            let buf = match extensions.get_mut::<FormattedFields<Self>>() {
+                Some(fields) => format!("{}{}", fields.fmt_fields, buf),
+                None => buf,
+            };
+            let fmt_fields = FormattedFields {
+                fmt_fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        thread_local! {
+            static BUF: RefCell<String> = RefCell::new(String::new());
+        }
+
+        BUF.with(|buf| {
+            let borrow = buf.try_borrow_mut();
+            let mut a;
+            let mut b;
+            let mut buf = match borrow {
+                Ok(buf) => {
+                    a = buf;
+                    &mut *a
+                }
+                _ => {
+                    b = String::new();
+                    &mut b
+                }
+            };
+
+            let ctx = self.make_ctx(ctx);
+            if self.fmt_event.format_event(&ctx, &mut buf, event).is_ok() {
+                let mut writer = self.make_writer.make_writer();
+                let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+            }
+
+            buf.clear();
+        });
+    }
+
+    unsafe fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
+        match () {
+            _ if id == TypeId::of::<Self>() => Some(self as *const Self as *const ()),
+            _ if id == TypeId::of::<E>() => Some(&self.fmt_event as *const E as *const ()),
+            _ if id == TypeId::of::<N>() => Some(&self.fmt_fields as *const N as *const ()),
+            _ => None,
+        }
+    }
+}
+
+/// Provides the current span context to a formatter.
+pub struct FmtContext<'a, S, N> {
+    pub(crate) ctx: Context<'a, S>,
+    pub(crate) fmt_fields: &'a N,
+}
+
+impl<'a, S, N> fmt::Debug for FmtContext<'a, S, N> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FmtContext").finish()
+    }
+}
+
+impl<'a, S, N> FormatFields<'a> for FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    fn format_fields<R: RecordFields>(
+        &self,
+        writer: &'a mut dyn fmt::Write,
+        fields: R,
+    ) -> fmt::Result {
+        self.fmt_fields.format_fields(writer, fields)
+    }
+}
+
+impl<'a, S, N> FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    /// Visits parent spans. Used to visit parent spans when formatting spans
+    /// and events
+    pub fn visit_spans<E, F>(&self, mut f: F) -> Result<(), E>
```

Comment:
since we aren't using the recursive approach anymore, this method could probably just be changed to return an iterator...code that calls it would then be much more idiomatic. On the other hand, if we don't want to _commit_ to changing this to use a smallvec rather than recursion,  which maybe we shouldn't do until we run some more benchmarks, it might be better to leave this as-is... what do you think?

Reply:
also, if we change this to return an iterator, we should probably move it to the more general `layer::Context` type...might be useful in cases other than formatting. actually...we could then still have a `visit_spans` method, but just have it use a `layer::Context` method that returns an iterator, this lets us have the nicer API in the more general place but _also_ hedges our bets so we can optimize this impl later if needed.

i can make that change if you like?

Reply:
I think:
- returning an iterator would be a good change, more ergonomic change.
- moving it to the more general `layer::Context` type would also be nice, alongside a yet-to-be-written `follows_from` iterator.

> i can make that change if you like?

Let me see how much bandwidth I have?

Reply:
So, I agree that this should become an iterator, but I'd much rather tackle this as part of a followup PR. Is that okay?

Reply:
> So, I agree that this should become an iterator, but I'd much rather tackle this as part of a followup PR. Is that okay?

that's fine, since it would be an additive change.